### ISSUE TITLE
Check replication from secondary point of view

### ISSUE THREAD

**Issue by Krysztophe** [OPMDG/check_pgactivity#363]
Following a discussion with @tobixen (https://github.com/OPMDG/check_pgactivity/pull/206#issuecomment-1832322517): he expressed the need of a service checking for the replication from the secondary point a view.

Just ideas and questions:

  * in which situations would this service be useful? (it was always assumed that the checks where done from the primary)
  * there are different cases:
    * pure streaming
    * streaming with log shipping (eg: the service could raise a WARNING if PG had to fall back on log shipping)
    * streaming ok but replay is paused
    * log shipping only, with or without archive_timeout
    * logical replication?

  * improve `is_hot_standby`
  * add a new service `is_streaming`?

**Comment by iovcho** [comment]
+1

**Comment by gleu** [comment]
To be honest, I don't know what to do with this. Care to explain what you need?

**Comment by Krysztophe** [comment]
> To be honest, I don't know what to do with this. Care to explain what you need?

Eg:
I have a secondary server in my department,
I do not have access to its primary (I have only the password for a replication role),
I want to be sure that the streaming is still going on.

Without much thought:
I imagine another service checking `pg_stat_walreceiver` the same way that check_streaming_delta uses `pg_stat_replication`.

(I  was wondering if making check_streaming_delta change its behaviour on a secondary would be possible,
but how to deal with a secondary which has itself replicas?)

**Comment by tobixen** [comment]
Blast from the past ... at the moment I'm not responsible for monitoring any postgres servers, so I don't need anything here, but if I remember correctly the plugin as it is can be run from the master and tell if there is a slave syncing from it, but it can not run on the slave and check that the slave is not far behind the master.  If I remember correct, to monitor the replication lag measured in seconds, it has to be done from the slave.  But my memory is not very good, so bear with me :-)

**Comment by gleu** [comment]
@Krysztophe I guess that last_msg_send_time could tell us when we've last heard from the primary. Would that be enough for you? (and my take would be to add another service, not hijacked an old one).

@tobixen Sorry, it took us quite a long time to get back on this. And thanks for the recollection!

**Comment by Krysztophe** [comment]
That sould be okay.

It is a bit weird to compute a duration here, while check_streaming_delta computes a data volume.
But pg_stat_walreceiver does not now where the primary really is (especially if disconnected…)
so let's go for a duration.

**Comment by gleu** [comment]
Yeah, there's no way to know the WAL location of the primary when you can only access a secondary. So, we'll got for duration, and a new service.

To sum it up, we need a new service, that errors out of connected to a primary, check the duration since the last pg_stat_walreceiver.last_msg_send_time, and compare it to a warning and/or critical value. I guess perfdata is the duration, rather than the date and time.
